-  challenge partition changes (allocation remove blobber)

ssardana08 opened the pull request:
…blobber

## Changes

## Fixes


## Tests 
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber:
- gosdk:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

ssardana08 committed:
- removing blobber and allocation from challenge partition on remove blobber

peterlimg commented on code/go/0chain.net/smartcontract/partitions/partitions.go:
```diff
@@ -80,6 +80,11 @@ func (p *Partitions) Size(state state.StateContextI) (int, error) {
 }
 
+// Remove removes the partition item from given partIndex in Partitions
+func (p *Partitions) Remove(state state.StateContextI, partIndex int, item PartitionItem) error {
```
Rename to `RemoveItem()` as this is for removing item from specific partition, and we'd better have the consistent APIs language, e.g. `GetItem()`, `UpdateItem`. The `Remove()` API may be used to remove specific partition in the future if needed.

We don't need`item` param, instead an `id string` is enough to locate the item.

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/models.go:
```diff
@@ -862,6 +862,54 @@ func (sa *StorageAllocation) removeBlobber(
+				if err != nil {
+					return nil, fmt.Errorf("error removing allocation from challenge partition: %v", err)
+				}
+				sa.BlobberDetails[i].ChallengePartitionLoc = nil
```
sa.BlobberDetail[i] has already been removed and replaced with the last item, so change it will cause unexpected error.

ssardana08 replied:
ahh yes! great catch @peterlimg

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/models.go:
```diff
@@ -862,6 +862,54 @@ func (sa *StorageAllocation) removeBlobber(
+					return nil, fmt.Errorf("error removing allocation from challenge partition: %v", err)
+				}
+				sa.BlobberDetails[i].ChallengePartitionLoc = nil
+				if err = blobberAllocChallPartition.Save(balances); err != nil {
```
We'd better call the `Save()` at the end of all the action in case any error happen, in this way we can avoid an unnecessary MPT writing.

ssardana08 replied:
if we don't save here, then Size() in next line will not return right size

ssardana08 replied:
I've made changes, please check if that looks fine

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/models.go:
```diff
@@ -862,6 +862,54 @@ func (sa *StorageAllocation) removeBlobber(
+						return nil, fmt.Errorf("error removing blobber from challenge partition: %v", err)
+					}
+
+					err = bcPartition.Save(balances)
```
Call the bcPartition.Save() out of the loop.

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/models.go:
```diff
@@ -862,6 +862,54 @@ func (sa *StorageAllocation) removeBlobber(
+						return nil, fmt.Errorf("error retrieving blobber challenge partition location: %v", err)
+					}
+
+					bcPartition, err := getBlobbersChallengeList(balances)
```
Get the bcPartition out of the loop

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/models.go:
```diff
@@ -862,6 +862,54 @@ func (sa *StorageAllocation) removeBlobber(
 			sa.BlobberDetails = sa.BlobberDetails[:len(sa.BlobberDetails)-1]
 			removedBlobber.Used -= d.Size
+
+			if d.ChallengePartitionLoc != nil {
```
Move the code block in the for loop for removing a blobber into a separate function or method to keep single method short, this is for keeping the method easy to read and maintain.

Also, we'd better wraps those partitions related actions in blobber's own methods, and call them to handle the business logic.

ssardana08 committed:
- challenge partition handling to a function
- bug on removing location of allocation blobber
- removeItem updated in partition

ssardana08 committed:
- test fix

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/models.go:
```diff
@@ -935,6 +942,58 @@ func (sa *StorageAllocation) changeBlobbers(
+	if err != nil {
+		return fmt.Errorf("error getting size of challenge partition: %v", err)
+	}
+	if blobberAllocChallSize-1 <= 0 {
```
The blobberAllocChallSize is the total item number in the partitions, so if all have been removed, it should be value of 0, not 1.

ssardana08 replied:
If we're doing Save at the end. It will be 1.. 
as it gets data from MPT and does size.
Since our partition is not saved yet. The size will be same as old one..

ssardana08 replied:
But in case if its 0 it should be 0. Yea you're right

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/models.go:
```diff
@@ -935,6 +942,58 @@ func (sa *StorageAllocation) changeBlobbers(
 	return blobbers, nil
 }
 
+func removeBlobberAllocationFromChallPartition(
```
Rename to removeBlobberAllocation() would be easier to read.

ssardana08 committed:
- pr comments

ssardana08 committed:
Merge branch 'staging' into blobber_challenge_enhancement

ssardana08 committed:
Merge branch 'staging' into blobber_challenge_enhancement

peterlimg reviewed (approved):
lgtm
